### ISSUE TITLE
TypeError: unhashable type: 'Stack' error raised when using Parallel adapter heads

### ISSUE THREAD

**Issue by hchoi-moveworks** [Adapter-Hub/adapter-transformers#221]
## Environment info

     
- `adapter-transformers` version:
- Platform: Linux
- Python version: 3.6.8
- PyTorch version (GPU?): GPU / 1.7
- Tensorflow version (GPU?): NA
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: Using nn.DataParallel

## Information

Model I am using (Bert, XLNet ...): 
BERT pretrained model with 3 custom adapters + heads are used.

Language I am using the model on (English, Chinese ...): EN

Adapter setup I am using (if any):
3 Adapters (with default configuration) and 3 Classification Head.

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
Multi-task finetuning using AdapterHub

Error below :
```
 (from logs) active head : [<bound method AdapterCompositionBlock.last of Stack[combined, resource_type, action]>]

Traceback (most recent call last):
  File "/home/hchoi/remote_sessions/.venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/hchoi/remote_sessions/.venv/lib/python3.6/site-packages/transformers/models/bert/modeling_bert.py", line 1092, in forward
    head_inputs, head_name=head, attention_mask=attention_mask, return_dict=return_dict, **kwargs
  File "/home/hchoi/remote_sessions/.venv/lib/python3.6/site-packages/transformers/adapters/heads.py", line 509, in forward_head
    if head not in self.heads:
  File "/home/hchoi/remote_sessions/.venv/lib/python3.6/site-packages/torch/nn/modules/container.py", line 304, in __contains__
    return key in self._modules
TypeError: unhashable type: 'Stack'
```

Modified code below
```

model = AutoModelWithHeads.from_pretrained('bert_base_uncased')

# 3 adapters and classification heads are added.
model.add_adapter('name_a')
model.add_classification_head('name_a',  {'num_labels' : 100})

model.add_adapter('name_b')
model.add_classification_head('name_b')

model.add_adapter('name_c')
model.add_classification_head('name_c',  {'num_labels' : 5})


# Use `Parallel` to enable multiple active heads.
adapter_names  = ['name_a', 'name_b', 'name_c']
model.active_heads =  ac.Parallel(adapter_names)

for name in adapter_names:
    model.train_adapter(name)
    
# Invoke forward pass. This will trigger the error. 
model(inputs)

```

## Expected behavior
Model forward pass should work.

**Comment by calpt** [comment]
Hi @hchoi-moveworks,

from you code snippet, it seems the error is caused by the line:
```
model.active_heads =  ac.Parallel(adapter_names)
```
`Parallel` expects the adapters as individual arguments. If a list is provided, it will try to parse the setup as a stack of adapters. Try to change this line to:
```
model.active_heads =  ac.Parallel(*adapter_names)
```
or
```
model.active_heads =  ac.Parallel('name_a', 'name_b', 'name_c')
```

**Comment by hchoi-moveworks** [comment]
Thank you @calpt ! The fix worked :) 

I have a follow up question. From the official adapterhub documentation, it seems like Parallel is only used for inference. 
Is there a recommended approach for using adapters for multi-task finetuning? (i.e. training models with **multiple** heads with adapter logic)

**Comment by JoPfeiff** [comment]
As far as I can tell, if you have task specific heads and adapters, it shouldn't make any difference if you train the adapters jointly or with the Parallel architecture. Each task has its own set of weights and thus there wouldn't be any gradient-wise interference.

There would only be an interaction if the tasks share a set of tuneable weights, i.e. if you stack a shared adapter on top of the task specific adapters. This however is an open research question 🙃
You would need to think about standard MT problems such as example sampling if you have uneven dataset sizes, but you could also address new ideas such as experimenting with in-batch task-distribution.
If this is something you are working on, I would be very curious about your results 😇

**Comment by hchoi-moveworks** [comment]
Thank you @JoPfeiff  for the response! You are precisely correct. My intention is to try out multi-head  + multi-task model using adapterhub models. 

I have 1 remaining question about backpropagation + gradients. 
>> it shouldn't make any difference if you train the adapters jointly or with the Parallel architecture. Each task has its own set of weights and thus there wouldn't be any gradient-wise interference.

For this, I thought that while gradients fully connected layers on classification heads are not shared, fully connected layers that are added in between the **transformer** layers in the adapter architecture would still be shared. Is this not the case? 

In summary I thought there were 3 types of weights. 
a) Frozen Bert Layers that includes Attention weights.
b) Weights added inside each transformer block (FF_up/FF_down) from adapterhub
c) FC layers added for each classification heads. 

While (a) is frozen and (c) does not share any gradients between tasks, gradients for (b) should be sharded between different heads right?

Please let me know if I'm missing anything!

**Comment by JoPfeiff** [comment]
b) being shared between tasks would be an implementation choice and would accumulate to just adding a single adapter which is used by every task.
They way you implemented it above introduces new adapter weights for every task 'name_a' 'name_b' and 'name_c', which in this setting, would not share any trainable weights.

The setting you described sounds more like you only want to add a single adapter, but multiple prediction heads.

**Comment by hchoi-moveworks** [comment]
>> The setting you described sounds more like you only want to add a single adapter, but multiple prediction heads.

Yup I think that works too as well. Is there a way to do this in adapterhub? From my trials, it seems like `add_classification_head()` API expects that `head_name` should **match** the `adapter_name`.

If I were to something like
```
model = AutoModelWithHeads.from_pretrained('bert_base_uncased')

# Add single adapter that will be shared for multi-classification heads.
model.add_adapter('my_adapter')

# Add multiple heads
model.add_classification_head('name_a',  {'num_labels' : 100})
model.add_classification_head('name_b')
model.add_classification_head('name_c',  {'num_labels' : 5})
```

Then I get the error 
```
 f"Model already contains a head with name '{head.name}'. Use overwrite_ok=True to force overwrite."
ValueError: Model already contains a head with name 'intent'. Use overwrite_ok=True to force overwrite.
```

Please let me know if there are any other ways/workarounds!

**Comment by calpt** [comment]
> From my trials, it seems like add_classification_head() API expects that head_name should match the adapter_name.

This shouldn't be the case, adapters and prediction heads can be added fully independently. The only case where having an adapter and a head with identical names matters is for activating adapters, e.g. `model.active_adapters = "my_adapter"`, as this will search for a prediction head with the same name and activate it together with the adapter. In all other cases, adapters and heads can be activated independently, using `active_adapters` and `active_head`.

If you do get an error, this might be a bug on our side. However, the code snippet you provided doesn't seem to produce the error you show. Could you check this again or provide some more detail?
